### PR TITLE
Introduce Ad hoc executor and auto-registration of opaque functions taking tensors

### PR DESCRIPTION
**Pull request by t-vi**
This also advances nemo using more of apex.

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
LGTM!

**Comment by crcrpar** [comment]
How would this PR change how traces would look?

**Comment by t-vi** [comment]
Not much except that the symbols are now on the AdHocExecutor and named slightly differently.
But so it tries to reduce the changes to the "global" environment, and I think things like properly running a "like" (meta) function will be correctness fixes once we look at dynamic shapes and such.

File: thunder/extend/__init__.py [hunk 1]
```diff
@@ -277,6 +280,85 @@ def register_implementation(
         self.implmap[_id] = impl
 
 
+class AdHocExecutor(OperatorExecutor):
+    """An "Anonymous" executor to be used for temporary registrations"""
```
[the remark this answers is not in the document]
Comment (t-vi) [reply thunder/extend/__init__.py:284]: The registrations are for a specific invocation of recipes / jit / interpreter.
We will use AdHoc executors for Recipe look-asides, too.

[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply thunder/extend/__init__.py:284]: How is it different from a regular OperatorExecutor? Should OperatorExecutor methods be extended to support what this AdHocExecutor does?

**Pull request merged**